src/engine/GroupBy.cpp

```diff
@@ -389,6 +412,162 @@ ProtoResult GroupBy::computeResult([[maybe_unused]] bool requestLaziness) {
+    processGroup<OUT_WIDTH>(aggregate, evaluationContext, 0, 0, &table, 0,
+                            aggregate._outCol, localVocab);
+  }
+  resultTable = std::move(table).toDynamic();
+}
+
+// _____________________________________________________________________________
+template <size_t IN_WIDTH, size_t OUT_WIDTH>
+cppcoro::generator<IdTable> GroupBy::computeResultLazily(
+    std::shared_ptr<const Result> subresult, std::vector<Aggregate> aggregates,
+    std::vector<HashMapAliasInformation> aggregateAliases,
+    std::vector<size_t> groupByCols, std::shared_ptr<LocalVocab> localVocab,
+    bool singleIdTable) const {
+  size_t inWidth = _subtree->getResultWidth();
+  AD_CONTRACT_CHECK(inWidth == IN_WIDTH || IN_WIDTH == 0);
+  LazyGroupBy lazyGroupBy{*localVocab, std::move(aggregateAliases),
+                          getExecutionContext()->getAllocator(),
+                          groupByCols.size()};
+
+  IdTable resultTable{getResultWidth(), getExecutionContext()->getAllocator()};
+
+  bool groupSplitAcrossTables = false;
+
+  std::vector<std::pair<size_t, Id>> currentGroupBlock;
+
+  for (IdTable& idTable : subresult->idTables()) {
+    if (idTable.empty()) {
+      continue;
+    }
+    AD_CORRECTNESS_CHECK(idTable.numColumns() == inWidth);
+    checkCancellation();
+
+    if (currentGroupBlock.empty()) {
+      for (size_t col : groupByCols) {
+        currentGroupBlock.emplace_back(col, idTable(0, col));
+      }
+    }
+
+    sparqlExpression::EvaluationContext evaluationContext =
+        createEvaluationContext(*localVocab, idTable);
+
+    size_t lastBlockStart = searchBlockBoundaries(
+        [&](size_t blockStart, size_t blockEnd) {
+          if (groupSplitAcrossTables) {
+            lazyGroupBy.processNextBlock(evaluationContext, blockStart,
+                                         blockEnd);
+            lazyGroupBy.commitRow(resultTable, evaluationContext,
+                                  currentGroupBlock, *this);
+            groupSplitAcrossTables = false;
+          } else {
+            // This processes the whole block in batches if possible
+            IdTableStatic<OUT_WIDTH> table =
+                std::move(resultTable).toStatic<OUT_WIDTH>();
+            processNextBlock<OUT_WIDTH>(table, aggregates, evaluationContext,
+                                        blockStart, blockEnd, localVocab.get(),
+                                        groupByCols);
+            resultTable = std::move(table).toDynamic();
+          }
+        },
+        idTable.asStaticView<IN_WIDTH>(), currentGroupBlock);
+    groupSplitAcrossTables = true;
+    lazyGroupBy.processNextBlock(evaluationContext, lastBlockStart,
+                                 idTable.size());
+    if (!singleIdTable && !resultTable.empty()) {
+      co_yield resultTable;
+      resultTable.clear();
+    }
+  }
+  // No need for final commit when loop was never entered.
+  if (!groupSplitAcrossTables) {
```

Comment:
I would use a separate variable for this (did we have any nonempty table etc.) Basically it is the `isFirst`variable I suggested above (maybe better `firstNonEmptyTableSeen` or something like this). This gives us the correct semantics here if I am not completely mistaken.